src/masternodes/loan.cpp

```diff
@@ -49,27 +63,35 @@ void CLoanView::ForEachLoanCollateralToken(std::function<bool (CollateralTokenKe
     ForEach<LoanSetCollateralTokenKey, CollateralTokenKey, uint256>(callback, start);
 }
 
-std::unique_ptr<CLoanView::CLoanSetCollateralTokenImpl> CLoanView::HasLoanCollateralToken(CollateralTokenKey const & key)
+std::optional<CLoanView::CLoanSetCollateralTokenImpl> CLoanView::HasLoanCollateralToken(CollateralTokenKey const & key)
 {
     auto it = LowerBound<LoanSetCollateralTokenKey>(key);
-    if (it.Valid() && it.Key().id == key.id)
+    if (it.Valid() && it.Key().id == key.id) {
         return GetLoanCollateralToken(it.Value());
+    } else if (const auto view = dynamic_cast<const CCustomCSView*>(this)) {
+        return GetCollateralTokenFromAttributes(*view, key.id);
+    }
```

Comment:
Make a pure virtual function in `CLoanView::GetCollateralTokenFromAttributes` then implement it in `CCustomCSView`.